lib/srv/db/sqlserver/auth.go

```diff
@@ -17,54 +17,123 @@ limitations under the License.
 package sqlserver
 
 import (
+	"context"
+	"crypto/x509"
+	"encoding/pem"
+	"strings"
+
 	"github.com/gravitational/trace"
 	"github.com/jcmturner/gokrb5/v8/client"
 	"github.com/jcmturner/gokrb5/v8/config"
+	"github.com/jcmturner/gokrb5/v8/credentials"
 	"github.com/jcmturner/gokrb5/v8/keytab"
 	"github.com/jcmturner/gokrb5/v8/spnego"
 
+	"github.com/gravitational/teleport/lib/auth"
 	"github.com/gravitational/teleport/lib/srv/db/common"
+	"github.com/gravitational/teleport/lib/srv/db/sqlserver/kinit"
 )
 
-// getAuth returns Kerberos authenticator used by SQL Server driver.
-//
-// TODO(r0mant): Unit-test this. In-memory Kerberos server?
-func (c *connector) getAuth(sessionCtx *common.Session) (*krbAuth, error) {
+type sqlClientI interface {
+	New() (*client.Client, error)
+}
+
+type keytabClient struct {
+	session *common.Session
+}
+
+// New returns a new keytabClient using a keytab file and then logging in with that keytab, which then obtains a TGT
+func (k *keytabClient) New() (*client.Client, error) {
 	// Load keytab.
-	keytab, err := keytab.Load(sessionCtx.Database.GetAD().KeytabFile)
+	kt, err := keytab.Load(k.session.Database.GetAD().KeytabFile)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	// Load krb5.conf.
-	config, err := config.Load(sessionCtx.Database.GetAD().Krb5File)
+	conf, err := config.Load(k.session.Database.GetAD().Krb5File)
 	if err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	// Create Kerberos client.
-	client := client.NewWithKeytab(
-		sessionCtx.DatabaseUser,
-		sessionCtx.Database.GetAD().Domain,
-		keytab,
-		config,
+	kbClient := client.NewWithKeytab(
+		k.session.DatabaseUser,
+		k.session.Database.GetAD().Domain,
+		kt,
+		conf,
 		// Active Directory does not commonly support FAST negotiation.
 		client.DisablePAFXFAST(true))
 
 	// Login.
-	err = client.Login()
+	err = kbClient.Login()
+	return kbClient, err
+}
+
+type kinitClient struct {
+	ctx     context.Context
+	session *common.Session
+	auth    auth.ClientI
+}
+
+// New returns a new kinitClient which configures and instantiates a new credentials cache which obtains a TGT
+func (c *kinitClient) New() (*client.Client, error) {
+	ldapPem, _ := pem.Decode([]byte(c.session.Database.GetAD().LDAPCert))
+
+	cert, err := x509.ParseCertificate(ldapPem.Bytes)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var k kinit.ProviderI = kinit.NewCommandProvider(
+		c.auth,
+		c.session.Identity.Username,
+		strings.ToUpper(c.session.Database.GetAD().Domain),
+		c.session.Database.GetAD().Domain,
+		c.session.Database.GetAD().Domain,
+		cert,
+	)
+
+	// create the kinit credentials cache using the previously prepared cert/key pair
+	err = k.CreateOrAppendCredentialsCache(c.ctx)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
```

Comment:
Does the CreateOrAppendCredentialsCache need to be always called if in case there is a valid ticket in the cache ?

Reply:
I guess I could rename the function? In this case when kinit runs again on the same cache, if the ticket is already valid, it is a no-op, so I could call it CreateOrUpdate instead?